### PR TITLE
Add refresh ability to user tokens, and expose tokens for storage

### PR DESCRIPTION
**Pull request by Dinnerbone**
This changes the API by adding an extra (optional) client secret field.

I also make some fields pub so that an application can retrieve them post-refresh and store them for the next run.

### CODE REVIEW

**Comment by Dinnerbone** [comment]
I can't seem to get the output of my `cargo fmt` to match what these tests require, sorry!

**Comment by Emilgardis** [comment]
wierd, could be a `rustfmt.toml` file somewhere in a subfolder

File: src/tokens/user_token.rs [hunk 1]
```diff
@@ -2,17 +2,21 @@ use crate::tokens::{
...
 use oauth2::{AccessToken, ClientId, RefreshToken};
 use oauth2::{HttpRequest, HttpResponse};
 use std::future::Future;
 
 /// An User Token from the [OAuth implicit code flow](https://dev.twitch.tv/docs/authentication/getting-tokens-oauth#oauth-implicit-code-flow) or [OAuth authorization code flow](https://dev.twitch.tv/docs/authentication/getting-tokens-oauth#oauth-authorization-code-flow)
 #[derive(Debug, Clone)]
 pub struct UserToken {
-    access_token: AccessToken,
+    /// The access token used to authenticate requests with
+    pub access_token: AccessToken,
```
Comment (Emilgardis) [inline src/tokens/user_token.rs:14]: This is already exposed with `TwitchToken::token()` which returns `&AccessToken`, however I can see the value in having this pub.

File: src/tokens/user_token.rs [hunk 2]
```diff
@@ -67,12 +75,28 @@ impl TwitchToken for UserToken {
...
+            {
+                crate::refresh_token(http_client, token, &self.client_id, &client_secret).await?
+            } else {
+                return Err(RefreshTokenError::NoRefreshToken);
+            };
+            self.access_token = access_token;
+            self.expires = expires;
+            self.refresh_token = refresh_token;
+        }
+        Ok(())
```
Comment (Emilgardis) [inline src/tokens/user_token.rs:99]: If no secret is available this should probably return a new error `RefreshTokenError::NoClientSecret`

**Review by Emilgardis — CHANGES_REQUESTED** [review]
Looks good!

**Comment by Emilgardis** [comment]
looks like a fix in rustfmt https://github.com/rust-lang/rustfmt/pull/4557

**Comment by Emilgardis** [comment]
thanks!

bors r+

**Pull request merged**